### PR TITLE
Remove compatibility with FOSRest < 2.1

### PR DESCRIPTION
**Pull request by jordisala1991**
I am targeting this branch, because this is BC.

## Changelog

```markdown
### Removed
- Removed compatibility with older versions of FOSRestBundle (<2.1)
```

## Subject

Since dropping a dependency is not a BC break, this is a good start, because we are doing a lot of code just to keep compat with FOSRest 1.

Those are the PR adding this code, to check if I did correctly or not: https://github.com/sonata-project/SonataMediaBundle/pull/1071 https://github.com/sonata-project/SonataMediaBundle/pull/1155 https://github.com/sonata-project/SonataMediaBundle/pull/1339

### CODE REVIEW

File: src/Controller/Api/MediaController.php [hunk 1]
```diff
@@ -406,24 +392,11 @@ protected function handleWriteMedium(Request $request, MediaInterface $media, Me
...
-
-                // NEXT_MAJOR: simplify when dropping FOSRest < 2.1
-                if (method_exists($context, 'disableMaxDepth')) {
-                    $context->disableMaxDepth();
-                } else {
-                    $context->setMaxDepth(0);
-                }
-                $view->setContext($context);
-            }
+            $context = new Context();
```
Comment (kunicmarko20) [inline src/Controller/Api/MediaController.php]: Is this even needed, can't it be replaced with `View` Annotation?

**Comment by kunicmarko20** [comment]
This looks more like a minor than a patch

File: src/Controller/Api/MediaController.php (hunk 1 above)
Comment (jordisala1991) [reply src/Controller/Api/MediaController.php]: I am not sure about that, It was originally implemented like that. Do you know a better way? I am not a user of FOSRest atm, just doing the cleanup

Comment (kunicmarko20) [reply src/Controller/Api/MediaController.php]: If I am correct, annotation [here](https://github.com/jordisala1991/SonataMediaBundle/blob/79952066c3f9dac4d227f05ac55dca943b196b44/src/Controller/Api/GalleryController.php#L85 )

`@View(serializerGroups={"sonata_api_read"}, serializerEnableMaxDepthChecks=true)`

Can replace:

```
            $view = FOSRestView::create($gallery);
            $context = new Context();
            $context->setGroups(['sonata_api_read']);
            $context->disableMaxDepth();
            $view->setContext($context);
            return $view;
```

to just 

```
return $gallery;
```

Only change `serializerEnableMaxDepthChecks` to `false`.

Comment (kunicmarko20) [reply src/Controller/Api/MediaController.php]: Hopefully, someone else is using FOSRest and could confirm this.

Comment (jordisala1991) [reply src/Controller/Api/MediaController.php]: But on our case we are doing this only on the `if`, on the else part we are returning a form

Comment (kunicmarko20) [reply src/Controller/Api/MediaController.php]: That does not affect form, form is not valid it returns validation errors, maybe I am using FOSRest wrong but I have it like that everywhere :smile:

Comment (jordisala1991) [reply src/Controller/Api/MediaController.php]: is this what you had in mind?

Comment (kunicmarko20) [reply src/Controller/Api/MediaController.php]: yes, but someone needs to test/confirm this will work.

File: src/Controller/Api/GalleryController.php [hunk 2]
```diff
@@ -412,6 +398,8 @@ public function deleteGalleryAction($id)
     /**
      * Write a GalleryHasMedia, this method is used by both POST and PUT action methods.
      *
+     * @View(serializerGroups={"sonata_api_read"}, serializerEnableMaxDepthChecks=false)
```
Comment (jordisala1991) [inline src/Controller/Api/GalleryController.php]: If this is the default value: `serializerEnableMaxDepthChecks=false` maybe we can remove it also?

**Comment by kunicmarko20** [comment]
Tests are failing because of change to View Annotation, will try to test this later today and confirm that we can go this way before you start fixing the tests.

**Comment by jordisala1991** [comment]
Okay, I will fix them tonight probably and I will create PR for all bundles using FOSRest

**Comment by kunicmarko20** [comment]
It works,

example (as in my comments):
https://github.com/kunicmarko20/fos-rest-test/blob/master/src/Controller/LabelController.php#L21

tests:
https://github.com/kunicmarko20/fos-rest-test/blob/master/tests/Controller/LabelControllerTest.php

travis passing:
https://travis-ci.org/kunicmarko20/fos-rest-test

The only thing I encountered that docs aren't good, so you could probably update that. [Here](https://sonata-project.org/bundles/media/3-x/doc/reference/api.html#setup) it says: 
```
sensio_framework_extra:
    view:    { annotations: false }
    router:  { annotations: true }
    request: { converters: true }
```

but view should be `true` and that is the default value so that part can even be removed. I also needed to add rule for `format_listener` [here](https://github.com/kunicmarko20/fos-rest-test/blob/master/config/packages/fos_rest.yaml#L5).

**Comment by jordisala1991** [comment]
Perfect, tests fixed too! :) Thank you for your time

File: src/Controller/Api/GalleryController.php [hunk 3]
```diff
@@ -520,6 +486,8 @@ protected function getMediaManager()
     /**
      * Write a Gallery, this method is used by both POST and PUT action methods.
      *
+     * @View(serializerGroups={"sonata_api_read"}, serializerEnableMaxDepthChecks=true)
```
Comment (jordisala1991) [inline src/Controller/Api/GalleryController.php]: Wait, @kunicmarko20  because this method is protected and called from the public methods, that's probably why it was created like that, I think I need to revert this change

Comment (jordisala1991) [reply src/Controller/Api/GalleryController.php]: I have decided to revert this, because it was used on a protected method, and will probably break BC that a method that returns a view, suddenly starts returning a media.

Let's not create any issue with this change. wdyt?

**Comment by jordisala1991** [comment]
If the code is clear enough to use the annotation, I will change to that ( like notificationbundle ). For cases like classification, here or pagebundle I will go for the code option

**Comment by kunicmarko20** [comment]
Well, it is just adding serialization group that is the same for every public method that calls that protected method, I am maybe missing something but that looks clear to me.

**Comment by jordisala1991** [comment]
I guess it is because I don't have the necessary knowledge about FOSRest

**Comment by kunicmarko20** [comment]
Ping @sonata-project/contributors, anyone using FOSRest and can help with this?

**Comment by jordisala1991** [comment]
I have doubts on when to use the annotation and when to use the php to create the view.
As far as I can see, the php code to create the view is only needed if you want to change the response code. Otherwise, you should always use annotations. Is that right @kunicmarko20 ?

**Comment by kunicmarko20** [comment]
You can change the status code in annotation also. Here is an example of how I use it with forms:

```
    /* *
     * @Rest\View(
     *  serializerGroups={
     *      "userComplete"
     *  },
     *  statusCode=Response::HTTP_CREATED
     * )
     */
    public function postAction(Request $request, UserService $userService)
    {
        $form = $this->createForm(UserType::class, $userDTO = new UserDTO());

        $form->submit($request->request->all());

        if ($form->isValid()) {
            return $userService->create($userDTO);
        }

        return $form;
    }
```

Using annotation looks cleaner to me than writing whole context and that stuff, but I do not want to block this so we can just leave it as is.

**Comment by jordisala1991** [comment]
Yes, you can change the code, but we have some cases where an `if` changes the status code. You can do that with annotations?

**Comment by kunicmarko20** [comment]
Don't think so but you can just do `return View::create($user, 201)` etc.

**Comment by jordisala1991** [comment]
We clarified a little bit how to deal with the api. Conclusions:

- Not all public methods needs to return a View. If there is no view, there is a listener on FOSRest that will wrap the object into a View.
- It is preferable to use annotations to define Views, unless you have to deal with special cases and status codes.
- We won’t change the protected methods to use annotations since it could be BC break. Maybe it could be done on master if it doesn’t conflict with point 2.
- In any case, why we want to use views with code or annotations if there is a listener that can deal with that? The reason is to define our custom serializedGroups when dealing with classes, and other methods like: `enableMaxDepth`.

This PR should be ready with all that in mind. Please review @sonata-project/contributors

**Comment by kunicmarko20** [comment]
Could you also update the docs per my [comment](https://github.com/sonata-project/SonataMediaBundle/pull/1389#issuecomment-365391737)

**Comment by jordisala1991** [comment]
Rebased, and docs fixed too! Should be good to go.

**Comment by jordisala1991** [comment]
Can we get this merged?

**Comment by greg0ire** [comment]
Sure
